File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generate(self, openCount, closeCount, N, curr, ans):
        if openCount == closeCount == N:
            ans.append(curr)
            return 
        
        if openCount < N:
            self.generate(openCount + 1, closeCount, N, curr + '(', ans)
        
        if closeCount < openCount:
            self.generate(openCount, closeCount + 1, N, curr + ')', ans)

    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.generate(0, 0, n, "",ans)
        return ans

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private void generate(int open, int close, int n, String cur, List<String> res) {
        if(open == close && open + close == 2*n){
            res.add(cur);
            return;
        }

        if(open < n){
            generate(open + 1, close, n, cur+'(', res);
        }

        if(close < open){
            generate(open, close + 1, n, cur+')', res);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        generate(0, 0, n, "", res);
        return res;
    }
}